File: fern/authentication/connections.py
import json

from composio.client.collections import ExpectedFieldInput
from composio.exceptions import NoItemsFound
from composio_openai import App, ComposioToolSet

toolset = ComposioToolSet()

integration_id = "fd6441eb-c51f-4153-9f31-d4de124ce3eb"
user_id = "00000000-0000-0000-0000-000000000002"
app = App.AGENCYZOOM


def connection_exists(user_id: str, app: App):
    entity = toolset.get_entity(user_id)
    try:
        connection = entity.get_connection(app=app)
        return connection.status == "ACTIVE"
    except NoItemsFound:
        return False


def request_connection_params(app: App, integration_id: str):
    return toolset.get_expected_params_for_user(integration_id=integration_id)


def prompt_for_connection_params(connection_params_info: list[ExpectedFieldInput]):
    print(
        "\nTo connect to this service, you'll need to provide the following authentication details:"
    )
    print("=" * 80)
    print()

    connection_params = {}
    for param in connection_params_info:
        param_dict = param.model_dump()
        print(f"Parameter: {param_dict['name']}")
        print(f"Type: {param_dict['type']}")
        print(f"Description: {param_dict['description']}")
        print("-" * 40)

        # Collect input from user
        value = input(f"Enter value for {param_dict['name']}: ")
        connection_params[param_dict["name"]] = value
        print()

    return connection_params


def initiate_connection(
    user_id: str, app: App, connection_params: dict, auth_mode: str, redirect_url: str
):
    entity = toolset.get_entity(user_id)
    connection = entity.initiate_connection(
        app_name=app,
        use_composio_auth=False,
        force_new_integration=True,
        connected_account_params=connection_params,
        auth_mode=auth_mode,
        redirect_url=redirect_url,
    )
    return connection


if connection_exists(user_id, app):
    print("Connection already exists for this user")
else:
    print("\nSetting up new connection...")
    connection_params_info = request_connection_params(app, integration_id)
    connection_params = prompt_for_connection_params(
        connection_params_info["expected_params"]
    )

    print(connection_params_info["expected_params"])

    # auth_schemes = toolset.get_auth_schemes(app=app)
    # for scheme in auth_schemes:
    #     print(scheme.model_dump_json(indent=4))

    # connection = initiate_connection(user_id, app, connection_params, "API_KEY", "https://www.google.com")
    # print(connection.model_dump_json(indent=4))

File: fern/authentication/connections.ts
import { OpenAIToolSet } from "composio-core";

const composioToolset = new OpenAIToolSet();

const user_id = "00000000-0000-0000-0000-000000000002";
const integration = await composioToolset.integrations.list({
  appName: "github",
});

enum App {
  AGENCYZOOM = "AGENCYZOOM",
  GMAIL = "GMAIL",
  // Add other apps as needed
}

interface ConnectionParams {
  [key: string]: string;
}

type AuthMode = "API_KEY" | "OAUTH2" | "OAUTH1" | "OAUTH1A" | "BASIC" | "BEARER_TOKEN" | 
  "GOOGLE_SERVICE_ACCOUNT" | "NO_AUTH" | "BASIC_WITH_JWT" | "COMPOSIO_LINK";

// Check if a connection exists for a user and app
async function connectionExists(userId: string, app: App): Promise<boolean> {
  try {
    const entity = await composioToolset.getEntity(userId);
    const connection = await entity.getConnection({ app });
    return connection.status === "ACTIVE";
  } catch (error) {
    // Check if the error is specifically about not finding a connection
    if (error instanceof Error && 
        (error.message?.includes("Could not find a connection") ||
        (error as any)?.errCode === "SDK::NO_CONNECTED_ACCOUNT_FOUND")) {
      console.log(`No active connection found for user ${userId} and app ${app}`);
      return false;
    }
    
    // Log other errors but still return false
    console.error("Error checking connection:", error);
    return false;
  }
}

// Get expected parameters for a connection
async function requestConnectionParams(app: string, authScheme: AuthMode) {
  return await composioToolset.apps.getRequiredParamsForAuthScheme({
    appId: app,
    authScheme: authScheme
  });
}

// Prompt for connection parameters (console version)
function getConnectionValues(connectionParamsInfo: any): ConnectionParams {
  console.log("\nTo connect to this service, you'll need to provide the following authentication details:");
  console.log("=".repeat(80));
  console.log();
  
  const connectionParams: ConnectionParams = {};
  
  // Handle required fields
  if (connectionParamsInfo.required_fields && connectionParamsInfo.required_fields.length > 0) {
    console.log("Required Parameters:");
    for (const paramName of connectionParamsInfo.required_fields) {
      console.log(`Parameter: ${paramName}`);
      console.log("-".repeat(40));
      
      // In browser environment, this would be replaced with form inputs
      const value = prompt(`Enter value for ${paramName}:`);
      if (value) {
        connectionParams[paramName] = value;
      }
      console.log();
    }
  }
  
  // Handle optional fields if needed
  if (connectionParamsInfo.optional_fields && connectionParamsInfo.optional_fields.length > 0) {
    console.log("Optional Parameters:");
    for (const paramName of connectionParamsInfo.optional_fields) {
      console.log(`Parameter: ${paramName}`);
      console.log("-".repeat(40));
      
      const value = prompt(`Enter value for ${paramName} (optional):`);
      if (value) {
        connectionParams[paramName] = value;
      }
      console.log();
    }
  }
  
  return connectionParams;
}

// Initiate a connection with the collected parameters
async function initiateConnection(
  userId: string,
  app: string,
  connectionParams: ConnectionParams,
  authMode: AuthMode,
  redirectUrl?: string
) {
  const entity = await composioToolset.getEntity(userId);
  const connection = await entity.initiateConnection({
    appName: app,
    authMode,
    connectionParams: connectionParams,
  });
  
  return connection;
}

// Main script logic
async function main() {
  const app = App.GMAIL;
  const authMode = "BEARER_TOKEN" as AuthMode;
  
  const exists = await connectionExists(user_id, app);
  
  if (exists) {
    console.log("Connection already exists for this user");
  } else {
    console.log("\nSetting up new connection...");
    const connectionParamsInfo = await requestConnectionParams(app, authMode);
    console.log(connectionParamsInfo);
    
    const connectionParams = getConnectionValues(connectionParamsInfo);
    
    console.log("Connection parameters collected:", connectionParams);
    
    if (Object.keys(connectionParams).length === 0) {
      console.log("No connection parameters provided. Exiting.");
      process.exit(0);
    }
    
    try {
      const connection = await initiateConnection(
        user_id,
        app,
        connectionParams,
        authMode,
      );
      console.log("Connection established successfully:");
      console.log(JSON.stringify(connection, null, 4));
    } catch (error) {
      console.error("Failed to establish connection:", error);
    }
  }
}

// Uncomment to run the script
main().catch(console.error);

